README
CUV Documentation

0.9.201107041204

Summary

CUV is a C++ template and Python library which makes it easy to use NVIDIA(tm)
CUDA.

Features

Supported Platforms:

  • This library was only tested on Ubuntu Karmic, Lucid and Maverick. It uses
    mostly standard components (except PyUBLAS) and should run without major
    modification on any current linux system.

Supported GPUs:

  • By default, code is generated for the lowest compute architecture. We
    recommend you change this to match your hardware. Using ccmake you can set
    the build variable "CUDA_ARCHITECTURE" for example to -arch=compute_20
  • All GT 9800 and GTX 280 and above
  • GT 9200 without convolutions. It might need some minor modifications to
    make the rest work. If you want to use that card and have problems, just
    get in contact.
  • On 8800GTS, random numbers and convolutions wont work.

Structure:

  • Like for example Matlab, CUV assumes that everything is an n-dimensional
    array called "tensor"
  • Tensors can have an arbitrary data-type and can be on the host (CPU-memory)
    or device (GPU-memory)
  • Tensors can be column-major or row-major (1-dimensional tensors are, by
    convention, row-major)
  • The library defines many functions which may or may not apply to all
    possible combinations. Variations are easy to add.
  • For convenience, we also wrap some of the functionality provided by Alex
    Krizhevsky on his website (http://www.cs.utoronto.ca/~kriz/) with
    permission. Thanks Alex for providing your code!

Python Integration

  • CUV plays well with python and numpy. That is, once you wrote your fast GPU
    functions in CUDA/C++, you can export them using Boost.Python. You can use
    Numpy for pre-processing and fancy stuff you have not yet implemented, then
    push the Numpy-matrix to the GPU, run your operations there, pull again to
    CPU and visualize using matplotlib. Great.

Implemented Functionality

  • Simple Linear Algebra for dense vectors and matrices (BLAS level 1,2,3)
  • Helpful functors and abstractions
  • Sparse matrices in DIA format and matrix-multiplication for these matrices
  • I/O functions using boost.serialization
  • Fast Random Number Generator
  • Up to now, CUV was used to build dense and sparse Neural Networks and
    Restricted Boltzmann Machines (RBM), convolutional or locally connected.

Documentation

  • Tutorials are available on http://www.ais.uni-bonn.de/~schulz/tag/cuv
  • The documentation can be generated from the code or accessed on the
    internet: http://www.ais.uni-bonn.de/deep_learning/doc/html/index.html

Contact

  • We are eager to help you getting started with CUV and improve the library
    continuously! If you have any questions, feel free to contact Hannes Schulz
    (schulz at ais dot uni-bonn dot de) or Andreas Mueller (amueller at ais dot
    uni-bonn dot de). You can find the website of our group at http://
    www.ais.uni-bonn.de/deep_learning/index.html.

Installation

Requirements

For C++ libs, you will need:

  • cmake (and cmake-curses-gui for easy configuration)
  • libboost-dev >= 1.37
  • libblas-dev
  • libtemplate-perl -- (we might get rid of this dependency soon)
  • NVIDIA CUDA (tm), including SDK. We support versions 3.X and 4.0
  • thrust library - included in CUDA since 4.0 (otherwise available from http:
    //code.google.com/p/thrust/)
  • doxygen (if you want to build the documentation yourself)

For Python Integration, you additionally have to install

  • pyublas -- from http://mathema.tician.de/software/pyublas
  • python-nose -- for python testing
  • python-dev

Optionally, install dependent libraries

  • cimg-dev for visualization of matrices (grayscale only, ATM)

Obtaining CUV

You should check out the git repository

   $ git clone git://github.com/deeplearningais/CUV.git

Installation Procedure

Building a debug version:

 $ cd cuv-version-source
 $ mkdir -p build/debug
 $ cd build/debug
 $ cmake -DCMAKE_BUILD_TYPE=Debug ../../
 $ ccmake .          # adjust paths to your system (cuda, thrust, pyublas, ...)!
                     # turn on/off optional libraries (CImg, ...)
 $ make -j
 $ ctest             # run tests to see if it went well
 $ sudo make install
 $ export PYTHONPATH=`pwd`/src      # only if you want python bindings

Building a release version:

 $ cd cuv-version-source
 $ mkdir -p build/release
 $ cd build/release
 $ cmake -DCMAKE_BUILD_TYPE=Release ../../
 $ ccmake .          # adjust paths to your system (cuda, thrust, pyublas, ...)!
                     # turn on/off optional libraries (CImg, ...)
 $ make -j
 $ ctest             # run tests to see if it went well
 $ sudo make install
 $ export PYTHONPATH=`pwd`/src      # only if you want python bindings

On Debian/Ubuntu systems, you can skip the sudo make install step and instead
do

 $ cpack -G DEB
 $ sudo dpkg -i cuv-VERSION.deb

Building the documentation

 $ cd build/debug    # change to the build directory
 $ make doc

Sample Code

We show two brief examples. For further inspiration, please take a look at the
test cases implemented in the src/tests directory.

Pushing and pulling of memory

C++ Code:

 #include <cuv.hpp>
     using namespace cuv;

     int main(void){
         tensor<float,host_memory_space> h(256);  // reserves space in host memory
         tensor<float,dev_memory_space>  d(256);  // reserves space in device memory

         fill(h,0);                          // terse form
         apply_0ary_functor(h,NF_FILL,0.f);    // more verbose

         d=h;                                // push to device
         sequence(d);                        // fill device vector with a sequence

         h=d;                                // pull to host
         for(int i=0;i<h.size();i++)
         {
             assert(d[i] == h[i]);
         }
     }

Python Code:

 import cuv_python as cp
 import numpy as np

 h = np.zeros((1,256))                                   # create numpy matrix
 d = cp.dev_tensor_float(h)                              # constructs by copying numpy_array

 h2 = np.zeros((1,256)).copy("F")                        # create numpy matrix
 d2 = cp.dev_tensor_float_cm(h2)                         # creates dev_tensor_float_cm (column-major float) object

 cp.fill(d,1)                                            # terse form
 cp.apply_nullary_functor(d,cp.nullary_functor.FILL,1)   # verbose form

 h = d.np                                                # pull and convert to numpy
 assert(np.sum(h) == 256)
 d.dealloc()                                             # explicitly deallocate memory (optional)

Simple Matrix operations

C++-Code

 #include <cuv.hpp>
 using namespace cuv;

 int main(void){
     tensor<float,dev_memory_space,column_major> C(2048,2048),A(2048,2048),B(2048,2048);

     fill(C,0);         // initialize to some defined value, not strictly necessary here
     sequence(A);
     sequence(B);

     apply_binary_functor(A,B,BF_MULT);  // elementwise multiplication
     A *= B;                             // operators also work (elementwise)
     prod(C,A,B, 'n','t');               // matrix multiplication
 }

Python Code

 import cuv_python as cp
 import numpy as np
 C = cp.dev_tensor_float_cm([2048,2048])   # column major tensor
 A = cp.dev_tensor_float_cm([2048,2048])
 B = cp.dev_tensor_float_cm([2048,2048])
 cp.fill(C,0)                       # fill with some defined values, not really necessary here
 cp.sequence(A)
 cp.sequence(B)
 cp.apply_binary_functor(B,A,cp.binary_functor.MULT) # elementwise multiplication
 B *= A                                              # operators also work (elementwise)
 cp.prod(C,A,B,'n','t')                              # matrix multiplication

The examples can be found in the "examples/" folder under "python" and "cpp"

Generated on Mon Jul 4 2011 12:04:53 for CUV by  doxygen 1.7.1


debian/changelog
cuv (0.97.45) raring; urgency=low

  [ Hannes Schulz ]
  * UNRELEASED

 -- Jenkins CI <jenkins@ais.uni-bonn.de>  Tue, 06 May 2014 11:51:38 +0200

cuv (0.97.44) raring; urgency=low

  ** SNAPSHOT build @533dc72b3ad138aa217bb4c96c5d86f5cc56869c **

  [ Nenad Biresev ]
  * FIX matrix_op_vec middle dim
  * FIX removed cout msg

  [ yi huang ]
  * FIX SF_COPY with mask

  [ Hannes Schulz ]
  * added debian files
  * RelWithDebInfo did not include NDEBUG. *ugh*.
  * UNRELEASED

 -- Hannes Schulz <schulz@ais.uni-bonn.de>  Mon, 25 Nov 2013 17:05:41 +0100

cuv (0.97.2-1ubuntu1~quantal2ubuntu1~5.gbpde2935) quantal; urgency=low

  ** SNAPSHOT build @de2935c59745df3c56f4e68dc7e66f188845516f **

  [ Nenad Biresev ]
  * FIX matrix_op_vec middle dim

 -- Hannes Schulz <schulz@ais.uni-bonn.de>  Mon, 26 Aug 2013 17:47:03 +0200

cuv (0.97.2-1ubuntu1~quantal2ubuntu1~4.gbpceeb2e) quantal; urgency=low

  ** SNAPSHOT build @ceeb2e1a5f84c2aac80d3f9972cc815adeb3c84c **

  [ Nenad Biresev ]
  * fixed function call for cuda mem copy
  * test for conv3d fixed
  * ENH bias added to border after theano convolution works
  * ENH subsampling in tuplewise op works
  * ENH tuplewise op mean added, and subsample
  * FIX assertions added for tensor_view
  * FIX warnings in cuda_ndarray
  * Fix warning in tensor.hpp
  * FIX warning about fprintf in cudandarray

  [ Hannes Schulz ]
  * force linking against cblas (weird ubuntu error)
  * FIX compatibility with cuda 5.5

  [ lukas koliogiannis ]
  * FIX bug fixed weightupdate in adagrad-kernel

  [ Hannes Schulz ]
  * FIX linking against python 3.2 if it is installed not desired
  * FIX library versioning
  * PKG add docs to all-target; install them
  * FIX make docs compilation optional again, fix install dir

  [ Nenad Biresev ]
  * FIX matrix_op_vec middle dim

  [ Hannes Schulz ]
  * added debian files

  [ Nenad Biresev ]
  * FIX matrix_op_vec middle dim

  [ Hannes Schulz ]
  * added debian files

 -- Hannes Schulz <schulz@ais.uni-bonn.de>  Wed, 14 Aug 2013 17:55:31 +0200

cuv (0.97.1-1ubuntu1~quantal2ubuntu1~2.gbp40eea5) quantal; urgency=low

  ** SNAPSHOT build @40eea55f621fc8bd0354cd35697ab489964495bc **

  [ Nenad Biresev ]
  * fixed function call for cuda mem copy
  * test for conv3d fixed
  * ENH bias added to border after theano convolution works
  * ENH subsampling in tuplewise op works
  * ENH tuplewise op mean added, and subsample
  * FIX assertions added for tensor_view
  * FIX warnings in cuda_ndarray
  * Fix warning in tensor.hpp

  [ Hannes Schulz ]
  * force linking against cblas (weird ubuntu error)
  * FIX compatibility with cuda 5.5

  [ lukas koliogiannis ]
  * FIX bug fixed weightupdate in adagrad-kernel

  [ Hannes Schulz ]
  * FIX linking against python 3.2 if it is installed not desired
  * FIX library versioning
  * PKG add docs to all-target; install them
  * FIX make docs compilation optional again, fix install dir
  * added debian files
  * bump to quantal
  * bump source version
  * packaging: compatibility with cuda 5.5
  * use cuda search path
  * added a shlibs.local file which explains where to find libcudart
  * debug package, api docs, shlibs.local
  * FIX debian package name
  * fix doc
  * fix changelog package name
  * docs again
  * extended description
  * specify fields in copyright template
  * ensure we have debug symbols
  * FIX warning that shlibs:depends is not defined for dbg package
  * more phonies
  * Build with debug info
  * docs again
  * update helper script
  * docs again
  * patch update
  * docs again
  * HACK fix CUDA_ARCH to "" for debugging build process
  * added copyright
  * control file: shorter extended description, versioning
  * turn off verbose makefiles
  * turn off testing using debian/rules (shouldn't need patch anymore now)
  * dependency cleanup
  * copyright again
  * description again
  * naming scheme again
  * dpkg build params makefile included
  * naming scheme again
  * dependencies
  * install file for libcuv0
  * docs package
  * docs package
  * docs package
  * patch update
  * FIX lintian error about numpy abi dependency
  * naming deps
  * added python-numpy dependency
  * FIX docs path
  * FIX docs path
  * FIX docs path
  * FIX numpy abi dep
  * FIX numpy abi dep
  * FIX numpy abi again
  * compile for all arch again
  * conflict with old cuv pkg

 -- Hannes Schulz <schulz@ais.uni-bonn.de>  Wed, 14 Aug 2013 13:54:23 +0200

cuv (0.97-1ubuntu1~quantal2ubuntu1~1.gbp17eb92) quantal; urgency=low

  ** SNAPSHOT build @17eb9245b54291f57c823c933d739b4a0afe4162 **

  [ Hannes Schulz ]
  * FIX 2d-copying of vectors
  * remove signed/unsigned warnings

  [ Nenad Biresev ]
  * modified separable filter
  * normalized box filter (hannes changed)
  * export hog commented out
  * removed tuplewise_spatial_op
  * added op for spatial norm
  * implemented separable filters with kernel param
  * test for separable filter
  * added theano ops flip_dim2and3 and shuffle_dim
  * deleted spatial op and implemented in cuvnet
  * flipping dimensions works
  * bug fixed in flipdim2and3
  * added test for theano ops
  * fixed the bug in reshape unsigned int
  * flipping is done for any dimension
  * parameter for flipping changed to cuv::extents
  * matrix_op_vec works for all axis
  * removed tuplewise_spatial_op
  * added eps param for the test tuplewise_op
  * modified documentation for tuplewise op, and deleted spatial_op
  * added eps param for tuplewise op, gradient fixed
  * removed copy to device in reshape
  * test for flip dimensions implemented shorter
  * head msg removed
  * added documentation for tuplewise_op_functor
  * resolved conflict, which was not resolved in last conflict
  * documentation for tuplewise_op_functor fixed

  [ Hannes Schulz ]
  * force linking against cblas (weird ubuntu error)
  * added debian files
  * bump to quantal

 -- Hannes Schulz <schulz@ais.uni-bonn.de>  Mon, 06 May 2013 11:42:13 +0200

cuv (0.96-1ubuntu1~quantal2) quantal; urgency=low


src/mainpage.hpp
//*LB*
// Copyright (c) 2010, University of Bonn, Institute for Computer Science VI
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
// 
//  * Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
//  * Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//  * Neither the name of the University of Bonn 
//    nor the names of its contributors may be used to endorse or promote
//    products derived from this software without specific prior written
//    permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//*LE*





/**
 *
 *
 * @mainpage
 *
 * @section summary  Summary
 *
 * CUV is a C++ template and Python library which makes it easy to use NVIDIA(tm) CUDA.
 *
 * @section features  Features
 *
 * Supported Platforms:
 * - This library was only tested on Ubuntu Karmic, Lucid and Maverick. It uses mostly standard
 *   components (except PyUBLAS) and should run without major modification on any current linux
 *   system.
 *
 * Supported GPUs:
 * - By default, code is generated for the lowest compute architecture.
 *   We recommend you change this to match your hardware.
 *   Using ccmake you can set the build variable "CUDA_ARCHITECTURE" for example to --arch=compute_20
 * - All GT 9800 and GTX 280 and above
 * - GT 9200 without convolutions. It might need some minor modifications to make the rest work.
 *   If you want to use that card and have problems, just get in contact.
 * - On 8800GTS, random numbers and convolutions wont work.
 *
 * Structure: 
 * - Like for example Matlab, CUV assumes that everything is an n-dimensional array called "tensor"
 * - Tensors can have an arbitrary data-type and can be on the host (CPU-memory) or device (GPU-memory)
 * - Tensors can be column-major or row-major (1-dimensional tensors are, by convention, row-major)
 * - The library defines many functions which may or may not apply to all possible combinations. Variations are easy to add.
 * - For convenience, we also wrap some of the functionality provided by Alex
 *   Krizhevsky on his website (http://www.cs.utoronto.ca/~kriz/) with
 *   permission. Thanks Alex for providing your code!
 *
 * Python Integration 
 * - CUV plays well with python and numpy. That is, once you wrote your fast
 *   GPU functions in CUDA/C++, you can export them using Boost.Python. You can
 *   use Numpy for pre-processing and fancy stuff you have not yet implemented,
 *   then push the Numpy-matrix to the GPU, run your operations there, pull
 *   again to CPU and visualize using matplotlib. Great.
 *
 * Implemented Functionality
 * - Simple Linear Algebra for dense vectors and matrices (BLAS level 1,2,3)
 * - Helpful functors and abstractions
 * - Sparse matrices in DIA format and matrix-multiplication for these matrices
 * - I/O functions using boost.serialization
 * - Fast Random Number Generator
 * - Up to now, CUV was used to build dense and sparse Neural Networks and
 *   Restricted Boltzmann Machines (RBM), convolutional or locally connected.
 * 
 * Documentation
 * - Tutorials are available on
 *   http://www.ais.uni-bonn.de/~schulz/tag/cuv
 * - The documentation can be generated from the code or accessed on the internet:
 *   http://www.ais.uni-bonn.de/deep_learning/doc/html/index.html
 *
 * Contact
 * - We are eager to help you getting started with CUV and improve the library continuously!
 *   If you have any questions, feel free to contact Hannes Schulz (schulz at ais dot uni-bonn dot de) or Andreas Mueller (amueller at ais dot uni-bonn dot de).
 *   You can find the website of our group at http://www.ais.uni-bonn.de/deep_learning/index.html.
 *
 * @section installation  Installation
 *
 * @subsection req  Requirements
 *
 * For C++ libs, you will need:
 * - cmake (and cmake-curses-gui for easy configuration)
 * - libboost-dev >= 1.37
 * - libblas-dev
 * - libtemplate-perl -- (we might get rid of this dependency soon)
 * - NVIDIA CUDA (tm), including SDK. We support versions 3.X and 4.0
 * - thrust library - included in CUDA since 4.0 (otherwise available from http://code.google.com/p/thrust/) 
 * - doxygen (if you want to build the documentation yourself)
 *
 * For Python Integration, you additionally have to install
 * - pyublas -- from http://mathema.tician.de/software/pyublas
 * - python-nose -- for python testing
 * - python-dev 
 *
 * Optionally, install dependent libraries
 * - cimg-dev for visualization of matrices (grayscale only, ATM)
 *
 *
 * @subsection obtaining  Obtaining CUV
 *
 * You should check out the git repository
 *   @code
 *   $ git clone git://github.com/deeplearningais/CUV.git
 *   @endcode
 *
 * @subsection instproc  Installation Procedure
 *
 * Building a debug version:
 *
 * @code
 * $ cd cuv-version-source
 * $ mkdir -p build/debug
 * $ cd build/debug
 * $ cmake -DCMAKE_BUILD_TYPE=Debug ../../
 * $ ccmake .          # adjust paths to your system (cuda, thrust, pyublas, ...)!
 *                     # turn on/off optional libraries (CImg, ...)
 * $ make -j
 * $ ctest             # run tests to see if it went well
 * $ sudo make install
 * $ export PYTHONPATH=`pwd`/src      # only if you want python bindings
 * @endcode
 *
 * Building a release version:
 *
 * @code
 * $ cd cuv-version-source
 * $ mkdir -p build/release
 * $ cd build/release
 * $ cmake -DCMAKE_BUILD_TYPE=Release ../../
 * $ ccmake .          # adjust paths to your system (cuda, thrust, pyublas, ...)!
 *                     # turn on/off optional libraries (CImg, ...)
 * $ make -j
 * $ ctest             # run tests to see if it went well
 * $ sudo make install
 * $ export PYTHONPATH=`pwd`/src      # only if you want python bindings
 * @endcode
 *
 * On Debian/Ubuntu systems, you can skip the sudo make install step and instead do
 *
 * @code
 * $ cpack -G DEB
 * $ sudo dpkg -i cuv-VERSION.deb
 * @endcode
 *
 * @subsection docinst  Building the documentation
 *
 * @code
 * $ cd build/debug    # change to the build directory
 * $ make doc
 * @endcode
 *
 * @section samplecode Sample Code
 *
 * We show two brief examples. For further inspiration, please take a look at
 * the test cases implemented in the @c src/tests  directory.
 *
 * @par Pushing and pulling of memory
 *
 * C++ Code:
 * @code
 * #include <cuv.hpp>
 *     using namespace cuv;
 * 
 *     int main(void){
 *         tensor<float,host_memory_space> h(256);  // reserves space in host memory
 *         tensor<float,dev_memory_space>  d(256);  // reserves space in device memory
 * 
 *         fill(h,0);                          // terse form
 *         apply_0ary_functor(h,NF_FILL,0.f);    // more verbose
 * 
 *         d=h;                                // push to device
 *         sequence(d);                        // fill device vector with a sequence
 * 
 *         h=d;                                // pull to host
 *         for(int i=0;i<h.size();i++)
 *         {
 *             assert(d[i] == h[i]);
 *         }
 *     }
 * @endcode
 *
 * Python Code:
 * @code
 * import cuv_python as cp
 * import numpy as np
 *
 * h = np.zeros((1,256))                                   # create numpy matrix
 * d = cp.dev_tensor_float(h)                              # constructs by copying numpy_array
 *
 * h2 = np.zeros((1,256)).copy("F")                        # create numpy matrix
 * d2 = cp.dev_tensor_float_cm(h2)                         # creates dev_tensor_float_cm (column-major float) object
 *
 * cp.fill(d,1)                                            # terse form
 * cp.apply_nullary_functor(d,cp.nullary_functor.FILL,1)   # verbose form
 *
 * h = d.np                                                # pull and convert to numpy
 * assert(np.sum(h) == 256)
 * d.dealloc()                                             # explicitly deallocate memory (optional)
 *
 * @endcode
 *
 * @par Simple Matrix operations
 *
 * C++-Code
 * @code
 *
 * #include <cuv.hpp>
 * using namespace cuv;
 * 
 * int main(void){
 *     tensor<float,dev_memory_space,column_major> C(2048,2048),A(2048,2048),B(2048,2048);
 * 
 *     fill(C,0);         // initialize to some defined value, not strictly necessary here
 *     sequence(A);
 *     sequence(B);
 * 
 *     apply_binary_functor(A,B,BF_MULT);  // elementwise multiplication
 *     A *= B;                             // operators also work (elementwise)
 *     prod(C,A,B, 'n','t');               // matrix multiplication
 * }
 * @endcode
 *
 * Python Code
 * @code
 * import cuv_python as cp
 * import numpy as np
 * C = cp.dev_tensor_float_cm([2048,2048])   # column major tensor
 * A = cp.dev_tensor_float_cm([2048,2048])   
 * B = cp.dev_tensor_float_cm([2048,2048])
 * cp.fill(C,0)                       # fill with some defined values, not really necessary here
 * cp.sequence(A)
 * cp.sequence(B)
 * cp.apply_binary_functor(B,A,cp.binary_functor.MULT) # elementwise multiplication
 * B *= A                                              # operators also work (elementwise)
 * cp.prod(C,A,B,'n','t')                              # matrix multiplication
 * @endcode
 *
 * The examples can be found in the "examples/" folder under "python" and "cpp"
 */


/** 
 * @namespace cuv
 * @brief contains all cuv functionality
 */

namespace cuv{
}

/**
 * @defgroup tools  Miscellaneous tools
 */
